rfcs/0007-handling-corruption.md

```diff
@@ -0,0 +1,324 @@
+non-transient corruption or a transient error (such as network connectivity
+issues).
+
+To support developers who value availability, we will introduce a new database
+option called `DbOptions.wal_auto_recovery`, similar in naming to
+`DbOptions.wal_enabled`. If set to `true`, then `DB::open_with_opts()` will
+perform WAL recovery asynchronously and automatically repair the WAL to the
+best of its ability, logging any corruption and repairs performed.
+
+To support the repair of the WAL, I propose introducing two new methods:
+`DB::wal_inspect()` and `DB::wal_repair()`. These methods will provide
+developers with tools to inspect the current contents of the WAL, including
+corrupted entries, and to repair the WAL to the best of its ability via entity
+recovery or truncation of the WAL.
+
+NOTE: `DbOptions.wal_auto_recovery` is a convenience for developers, as the
+same result can be achieved by calling `DB::open_with_opts()`, immediately
+calling `DB::wal_repair()`, and then calling `DB::open_with_opts()` again.
+
+### Handling SST Corruption
+Corruption of the SST can occur at any stage before, during, or after a
+database operation. Therefore, the most suitable place to report corruption is
+during the `get` call to retrieve keys from the database. If a key is NOT found
+and SlateDB detects corruption while searching for the key, an error should be
+reported. However, if SlateDB encounters corruption while searching for AND
+finding a key, it should not return an error. This ensures the database remains
+available even in the presence of corruption.
+
+The error returned should be a distinct type which can easily be matched to
+determine if the error is due to non transient corruption or a transient error.
+(network connectivity, etc..)
+
+The rationale behind this approach is rooted in the assumption that SlateDB's
+design is a manifestation of a physical system. Linux system administrators
+would not use physical disks that refuse access to all usable sectors due to a
+single bad sector. Such a disk would be impractical in many LVM (Logical Volume
+Manager) or RAID configurations. Similarly, the ability for developers who wish
+to use SlateDB to achieve Logical Durability is severely compromised if access
+to uncorrupted data blocks is not allowed.
+
+To support those operators who value durability, the developer making the call
+to `get` can decide to abort the application if and corruption error is returned.
+The application can then provide operators with detailed information included
+in the error returned by SlateDB on the nature of the corruption and suggestions
+on how to fix it.
+
+To support those operators who value availability, the error can be logged or
+ignored until either the database is deleted, or corruption repaired.
+
+To support repair of the SST, I propose we introduce new methods called
+`DB::sst_check()` , `DB::sst_inspect()` and `DB::sst_repair()` providing
```

Comment:
how about a single `repair()` method for all of these? 

it seems rocksdb only offered one operation for this:

https://docs.rs/rocksdb/latest/rocksdb/struct.DBCommon.html#method.repair

imo SST/WAL is kinda internal details for users, if user got into data corruption, a single `repair()` method with some options might be easier to document and easier to become the goto place for resolve these issues.

Reply:
This is a great point. In my mind there are two possible sources of corruption, WAL and SST. However, for most users who just want to get up and running after encountering corruption, a simple "fix everything" is quite seductive. The downside to the "fix everything" is there are may be many terabytes of SST files taking some unbounded amount of time to process, when only the WAL contains the corruption. 

The thing I'm most worried about is that they may run the "repair" in haste, losing data which *might* have been recoverable.

The more I think about this... both the WAL and SST repair operation should make copies of the corrupted files for inspection by operators after recovery. This provides them with both an avenue to "fix everything" while also allowing both operators and potentially developers or even upstream developers access to the original corruption for forensic analysis.

My background is in SaaS where the impetus is to resume operation ASAP, and preform RCA and data validation afterwards. (I've never worked at a financial institution, so take that into consideration)

Reply:
IMO, a repair() that takes an SST ID (either WAL or compacted) would be a good way to go about this.

And, yes, putting the bad SST in "corrupt/" folder along with any debug data would be quite nice.